gsms.py

```diff
@@ -153,6 +157,176 @@ def get_win_path(folder_id: str) -> str:
     # Free memory used by pointer
     _CoTaskMemFree(path_pointer)
 
+    print(path)
+
+    return path
+
+
+def copy_image(src_image: str, dst_image: str) -> None:
+    """
+    Copies an image from src_image to dst_image if the dst is empty or different
+
+    Parameters
+    ----------
+    src_image: str
+        Source path of the image
+    dst_image: str
+        Destination path of the image
+
+    Returns
+    -------
+    None
+
+    Examples
+    --------
+    >>> copy_image("C:\\Image1.png", "D:\\Image1.png")
+    """
+    # if src_image exists and dst_image does not exist
+    if os.path.isfile(src_image) and not os.path.isfile(dst_image):
+        shutil.copy2(src=src_image, dst=dst_image)  # copy2 preserves metadata
+        print(f'Copied box-art image to: {dst_image}')
+    else:
+        print(f'No box-art image found at: {src_image}')
+
+
+def has_app(sunshine_apps: any, name: str) -> bool:
+    """
+    Checks if a given app name is in the sunshine_apps object
+
+    Parameters
+    ----------
+    sunshine_apps: any
+        Parsed JSON object of the sunshine `apps.json`
+    name: str
+        Name to check for
+
+    Returns
+    -------
+    bool:
+        True if the app is in the sunshine_apps object. Otherwise False
+
+    Examples
+    --------
+    >>> has_app(any, "Game Name")
+    False
+    """
+    app_exists = False
+
+    for existing_app in sunshine_apps['apps']:
+        if name == existing_app['name']:
+            app_exists = True
+            print(f'{name} app already exist in Sunshine apps.json, skipping.')
+            break
+
+    return app_exists
+
+
+def add_game(sunshine_apps: any, name: str, logfile: str, cmd: str, working_dir: str, image_path: str) -> None:
+    """
+    Function to add an app to the sunshine_apps object passed in
+
+    Parameters
+    ----------
+    sunshine_apps: any
+        Parsed JSON object of the sunshine `apps.json`
+    name: str
+        Name of the app
+    logfile: str
+        Logfile name for the app
+    cmd: str
+        Commandline to start the app
+    working_dir: str
+        Working directory for the app
+    image_path: str
+        Path to an image file to display as box-art
+
+    Returns
+    -------
+    None
+
+    Examples
+    --------
+    >>> add_game(any, "Game Name", "game.log", "game.exe", "C:\\gamedir", "C:\\gamedir\\image.png")
+    """
+
+    working_dir = known_path_to_absolute(working_dir)
+    cmd = known_path_to_absolute(cmd)
+
+    # remove final path separator but only if it exists
+    while working_dir.endswith(os.sep):
+        working_dir = working_dir[:-1]
+
+    # remove preceding separator on the command if it exists
+    while cmd.startswith(os.sep):
+        cmd = cmd[1:]
+
+    # we don't need ot keep quotes around the path or working directory
+    working_dir = working_dir.replace('"', "")
+    cmd = cmd.replace('"', '')
+
+    # steam URI commands need some special treatment
```

Comment:
Probably any commands that starts with `start` should be a detached command? As far as I know these will never work in Sunshine as a normal `cmd`.

Also, as far as I know, any URI would also be required to be a detached command.

So maybe this would be more simple/shorter:
```py
# process detached commands
detached = False

# cmd begins with "start", this must be a detached command
if cmd.lower().startswith("start"):
    detached = True
    cmd = cmd[5:].strip()

# command is a URI, this must be a detached command
if '://' in cmd:
    detached = True

sunshine_apps['apps'].append(
    {
        'name': name,
        'output': logfile,
        'detached': [cmd] if detached else None,
        'cmd': cmd if not detached else None,
        'working-dir': working_dir,
        'image-path': image_path
    }

# remove the empty keys
for k, v in sunshine_apps['apps'][-1].items()
    if v is None:
        del sunshine_apps['apps'][-1][k]
```

Reply:
the last loop to remove the not filled once complains about the dict changing size and thus errors and aborts the script.

Reply:
I have found a work around by using this

```py
detached = False

# cmd begins with "start", this must be a detached command
if cmd.lower().startswith("start"):
    detached = True
    cmd = cmd[5:].strip()

# command is a URI, this must be a detached command
if '://' in cmd:
    detached = True

    # if the URI uses the steam protocol we prepend the steam executeable
    if 'steam://' in cmd:
        cmd = f"steam {cmd}"
# if it's not a URI we check if the command includes the working_directory
elif not cmd.startswith(working_dir):
    cmd = os.path.join(working_dir, cmd)  

data = {
    'name': name,
    'output': logfile,
    'working-dir': working_dir,
    'image-path': image_path
}

# we add the command to the appropriate field
if detached:
    data['detached'] = [cmd]
else:
    data['cmd'] = cmd

sunshine_apps['apps'].append(data)
```